gtk/src/toga_gtk/hardware/location.py

```diff
@@ -0,0 +1,146 @@
+from enum import IntEnum, auto
+
+import gi
+
+from toga import LatLng
+from toga_gtk.libs import Gio, GLib, GObject
+
+gi.require_version("Geoclue", "2.0")
+from gi.repository import Geoclue  # noqa: E402
+
+
+def toga_location(location):
+    """Convert a ``Geoclue.Location`` into ``OnLocationChangeHandler`` kwargs."""
+    latlng = LatLng(*location.get_properties("latitude", "longitude"))
+    altitude = location.get_property("altitude")
+
+    return {
+        "location": latlng,
+        "altitude": altitude,
+    }
+
+
+class States(IntEnum):
+    STARTING = auto()
+    """Waiting for the Geoclue client to initialise."""
+
+    READY = auto()
+    """Geoclue is ready to connect."""
+
+    MONITORING = auto()
+    """Actively monitoring the location."""
+
+    FAILED = auto()
+    """``Geoclue`` was unable to retrieve the location due to a generic error."""
+
+    DENIED = auto()
+    """Permission to read the location was denied."""
+
+
+class Location(GObject.Object):
+    # Implement as GObject.Object to allow notifying on state changes within the event loop
+    state = GObject.Property(type=int, default=States.STARTING)
+
+    def __init__(self, interface):
+        super().__init__()
+
+        self.interface = interface
+
+        self.geoclue_simple = None
+        self.notify_location_handle = None
+        self.notify_active_listener = None
+        self.in_flatpak = None
+
+        def new_finish(_, async_result):
+            try:
+                self.geoclue_simple = Geoclue.Simple.new_finish(async_result)
+            except GLib.Error as e:
+                if e.matches(Gio.DBusError, Gio.DBusError.ACCESS_DENIED):
+                    # TODO: In practice, I cannot get this to occur. Maybe it is KDE's portal
+                    # implementation, but when I deny location permissions in the prompt,
+                    # the failure is a generic DBusError.FAILED, not a permission denied
+                    # failure.
+                    self.props.state = States.DENIED
```

Comment:
It isn't KDE, the behaviour is the same in GNOME. I'll spend a bit more time poking around to see if there's a way to determine whether location permissions are available. I also need to see what happens when Geoclue isn't available at all. I suppose the import wouldn't work, and a separate kind of `NotImplementedError` should be raised in the construction in that case?

Reply:
Done in https://github.com/beeware/toga/pull/2999/commits/0e0ec80c4e03146e3f4678c29852c2ba48017cf8#diff-7d2a7a8b2296cd8aa08963e1fb6e2e9420f1051f910fb4b820db6348de4f2d3eR67